### PR TITLE
libnetwork/overlay:fix sandbox deadlock

### PR DESCRIPTION
**Pull request by evol262**
This PR updates #42842, but I don't have rights to their repo. @dojci, I've had some conversations with a maintainer, and moving to a goroutine instead should finally get this merged. Want to pull it in?

> We are using docker swarm with overlay network driver in the environment where stacks are very often updated. There was a problem several times a week when the docker daemon unexpectedly hangs forever, the docker daemon was unable to process update commands like remove and deploy stacks/services as well as container operations. The only thing that helped was the force-killing docker daemon and start it again. This problem often happens when we want to redeploy stack/service. When the docker gets stuck, the service from the given stack always remains in the Remove state forever. It is very difficult to reproduce this problem and we didn't find a reliable way to trigger this deadlock issue. We could only wait for it to happen spontaneously (2-4 times a week).

**- What I did**
 I changed the way how the peerInit operation is called from joinSandbox() to avoid deadlock.
 
**- How I did it**
Spooling `initSandboxPeerDB` off into a goroutine so if it needs to write back to `peerOpCh`, we do not have the same goroutine both producing and consuming, deadlocking the channel.

**- Description for the changelog**

Operations performed on overlay network sandboxes are handled by
dispatching operations send through a channel. This allows for
asynchronous operations to be performed which, since they are
not called from within another function, are able to operate in
an idempotent manner with a known/measurable starting state from
which an identical series of iterative actions can be performed.

However, it was possible in some cases for an operation dispatched
from this channel to write a message back to the channel in the
case of joining a network when a sufficient volume of sandboxes
were operated on.

A goroutine which is simultaneously reading and writing to an
unbuffered channel can deadlock if it sends a message to a channel
then waits for it to be consumed and completed, since the only
available goroutine is more or less "talking to itself". In order
to break this deadlock, in the observed race, a goroutine is now
created to send the message to the channel.

### CODE REVIEW

File: libnetwork/drivers/overlay/peerdb.go [hunk 1]
```diff
@@ -246,8 +246,7 @@ func (d *driver) peerDbDelete(nid, eid string, peerIP net.IP, peerIPMask net.IPM
 // networkDB has already delivered some events of peers already available on remote nodes,
 // these peers are saved into the peerDB and this function is used to properly configure
 // the network sandbox with all those peers that got previously notified.
-// Note also that this method sends a single message on the channel and the go routine on the
-// other side, will atomically loop on the whole table of peers and will program their state
+// Note also that this method will atomically loop on the whole table of peers and will program their state
```
Comment (corhere) [inline libnetwork/drivers/overlay/peerdb.go:249]: Please revert the changes to this comment now that the implementation has also been reverted

**Review by corhere — COMMENTED** [review]
Please rebase and squash down to a single commit. You can attribute the change to @dojci by adding [a `Co-authored-by` trailer](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) to the commit message if the original authorship information is clobbered by squashing.

**Comment by evol262** [comment]
> @evol262 could you address @corhere's comment and squash the commits? I think after that we should be able to move this forward

Absolutely. I spoke to him yesterday. Will squash today

**Comment by thaJeztah** [comment]
I think something went wrong with your rebase; looks like your change to use a go routine went AWOL 😅

**Review by thaJeztah — APPROVED** [review]
code LGTM

would perhaps be nice to include a description of the change in the commit message, e.g. the part you wrote in the top comment on GitHub (but could include some of the problem description);

> Spooling initSandboxPeerDB off into a goroutine so if it needs to write back to peerOpCh, we do not have the same goroutine both producing and consuming, deadlocking the channel.

**Comment by evol262** [comment]
> code LGTM
> 
> would perhaps be nice to include a description of the change in the commit message, e.g. the part you wrote in the top comment on GitHub (but could include some of the problem description);
> 
> > Spooling initSandboxPeerDB off into a goroutine so if it needs to write back to peerOpCh, we do not have the same goroutine both producing and consuming, deadlocking the channel.

Done, in both the "proper" commit message and the "summary for the changelog" here.

**Review by corhere — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by tianon — APPROVED** [review]
